contracts/exchangeIssuance/FlashMintDex.sol

```diff
@@ -0,0 +1,612 @@
+     */
+    function getRedeemExactSet(
+        IssueRedeemParams memory _redeemParams,
+        DEXAdapterV2.SwapData memory _swapDataWethToOutputToken
+    )
+        external
+        returns (uint256)
+    {
+        uint256 totalWethReceived = 0;
+        (,, uint256[] memory wethReceived) = _getWethReceivedPerComponent(_redeemParams);
+
+        for (uint256 i = 0; i < wethReceived.length; i++) {
+            totalWethReceived += wethReceived[i];
+        }
+        return dexAdapter.getAmountOut(_swapDataWethToOutputToken, totalWethReceived);
+    }
+
+    /**
+    * Issues an exact amount of SetTokens for given amount of ETH.
+    * Any leftover ether is returned to the caller.
+    *
+    * @param _issueParams           Struct containing addresses, amounts, and swap data for issuance
+    *
+    * @return ethSpent              Amount of ETH spent
+    */
+    function issueExactSetFromETH(IssueRedeemParams memory _issueParams)
+        external
+        payable
+        isValidModuleAndSet(_issueParams.issuanceModule, address(_issueParams.setToken))
+        nonReentrant
+        returns (uint256 ethSpent)
+    {
+        require(msg.value > 0, "FlashMint: NO ETH SENT");
+
+        IWETH(WETH).deposit{value: msg.value}();
+
+        uint256 ethUsedForIssuance = _issueExactSetFromWeth(_issueParams);
+
+        require(ethUsedForIssuance <= msg.value, "FlashMint: OVERSPENT ETH");
+
+        uint256 ethReturned = msg.value.sub(ethUsedForIssuance);
+        if (ethReturned > 0) {
+            IWETH(WETH).withdraw(ethReturned);
+            payable(msg.sender).sendValue(ethReturned);
+        }
+        ethSpent = msg.value.sub(ethReturned);
+
+        emit FlashMint(msg.sender, _issueParams.setToken, IERC20(ETH_ADDRESS), ethSpent, _issueParams.amountSetToken);
+    }
+
+    /**
+    * Issues an exact amount of SetTokens for given amount of input ERC20 tokens.
+    * Any leftover value is swapped back to the payment token and returned to the caller.
+    *
+    * @param _issueParams           Struct containing addresses, amounts, and swap data for issuance
+    * @param _paymentInfo           Struct containing input token address, max amount to spend, and swap data to trade for WETH
+    *
+    * @return paymentTokenSpent     Amount of input token spent
+    */
+    function issueExactSetFromERC20(IssueRedeemParams memory _issueParams, PaymentInfo memory _paymentInfo)
+        external
+        isValidModuleAndSet(_issueParams.issuanceModule, address(_issueParams.setToken))
+        nonReentrant
+        returns (uint256 paymentTokenSpent)
+    {
+        _paymentInfo.token.safeTransferFrom(msg.sender, address(this), _paymentInfo.limitAmt);
+        uint256 wethReceived = _swapPaymentTokenForWeth(_paymentInfo.token, _paymentInfo.limitAmt, _paymentInfo.swapDataTokenToWeth);
+
+        uint256 wethSpent = _issueExactSetFromWeth(_issueParams);
+        require(wethSpent <= wethReceived, "FlashMint: OVERSPENT WETH");
```

Comment:
repeat do we need this check? especially in relation to 294 and 297

Reply:
@pblivin0x Reasons for leaving it in:
1) Makes it extra clear in code that slippage is enforced
2) It's a nicer error message vs. the more general SafeMath Underflow we'd get in line 294.

I can take it out though, if you don't think that's helpful!

Reply:
Left this check in, and added a test for it in [dc7a943](https://github.com/IndexCoop/index-coop-smart-contracts/pull/181/commits/dc7a9432898a6771d2324cd6ca00307d97dc51ff)